### PR TITLE
[mqtt.homeassistant] Improve support for Lock component

### PR DESCRIPTION
**Pull request by ccutrer**
* handle state and command payloads differing (as they do by default)
 * expose full state possibilities and OPEN command by adding a TextValue channel

### CODE REVIEW

**Review by antroids — APPROVED** [review]
LGTM, 
Thank you!

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java [hunk 1]
```diff
@@ -37,30 +38,61 @@
...
 
     /**
-     * Create a string value with a limited number of allowed states.
+     * Create a string value with a limited number of allowed states and commands.
      *
      * @param states Allowed states. Empty states are filtered out. If the resulting set is empty, all string values
      *            will be allowed.
+     * @param commands Allowed commands. Empty commands are filtered out. If the resulting set is empty, all string
+     *            values
+     *            will be allowed.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java:50]:
```suggestion
     *            values will be allowed.
```

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java [hunk 2]
```diff
@@ -37,30 +38,61 @@
...
-    public TextValue(String[] states) {
+    public TextValue(String[] states, String[] commands) {
         super(CoreItemFactory.STRING, List.of(StringType.class));
         Set<String> s = Stream.of(states).filter(not(String::isBlank)).collect(Collectors.toSet());
         if (!s.isEmpty()) {
             this.states = s;
         } else {
             this.states = null;
         }
+        s = Stream.of(commands).filter(not(String::isBlank)).collect(Collectors.toSet());
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java:60]: Maybe better call this variable `c` or even `commands`?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java [hunk 3]
```diff
@@ -37,30 +38,61 @@
...
+        if (!s.isEmpty()) {
+            this.commands = s;
+        } else {
+            this.commands = this.states;
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java:64]: Not all states are always automatically allowed commands. If no commands are provided, why not simply leave the commands empty here?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java [hunk 4]
```diff
@@ -37,30 +38,61 @@
...
+        }
+    }
+
+    /**
+     * Create a string value with a limited number of allowed states.
+     *
+     * @param states Allowed states. Empty states are filtered out. If the resulting set is empty, all string values
+     *            will be allowed.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/TextValue.java:72]: Maybe also mention that it is assumed that all states are also valid commands and that if this is not the case the other constructor should be used.

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Lock.java [hunk 5]
```diff
@@ -39,30 +46,99 @@ static class ChannelConfiguration extends AbstractChannelConfiguration {
...
     public Lock(ComponentFactory.ComponentConfiguration componentConfiguration) {
         super(componentConfiguration, ChannelConfiguration.class);
 
-        // We do not support all HomeAssistant quirks
-        if (channelConfiguration.optimistic && !channelConfiguration.stateTopic.isBlank()) {
-            throw new ConfigurationException("Component:Lock does not support forced optimistic mode");
-        }
+        this.optimistic = channelConfiguration.optimistic || channelConfiguration.stateTopic.isBlank();
+
+        lockValue = new OnOffValue(new String[] { channelConfiguration.stateLocked },
```
Comment (ccutrer) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Lock.java:80]: I could see an argument that UNLOCKING should be considered locked for this binary channel, but for now I took the conservative approach that anything that's not locked should be considered unlocked.

**Pull request merged**